modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/SortNode.java

```diff
@@ -38,16 +43,45 @@
     /** Rows buffer. */
     private final PriorityQueue<RowT> rows;
 
+    /** SQL select limit. Negative if disabled. */
+    private final int limit;
+
+    /** Reverse-ordered rows in case of limited sort. */
+    private List<RowT> reversed;
```

Comment:
```suggestion
    /** Reverse-ordered rows in case of limited sort. */
    private List<RowT> stack;
```
Also, I'd use fastutil Stack interface here, or even add Stack backed by ArrayList to core module.

Reply:
Also, I think we can reuse the same array for both structures: PriorityQueue and Stack.

E.g. using ObjectHeapPriorityQueue and write trivial Stack implementation.
Actually, PriorityQueue stores data in heap structure that is partially sorted.
It looks easy to implement heap-sort algorithm to get the array sorted instantly.

Reply:
I think most optimizations and refactoring should be done separately. Here just port from AI2 with fixes real bugs